### PR TITLE
feat: Add rule no-constant-binary-expression

### PR DESCRIPTION
**Pull request by captbaritone**
#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[X] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))

**Please describe what the rule should do:**

**What category of rule is this? (place an "X" next to just one item)**

[X] Warns about a potential error

**Provide 2-3 code examples that this rule will warn about:**

```js
value === (0 || undefined || null)

!value !== null

someFunction() != null ?? false

x + b ?? 0
```

**Why should this rule be included in ESLint (instead of a plugin)?**

Recommended by TSC in #13752.

#### What changes did you make? (Give an overview)

Adds a rule which is similar in nature to `no-constant-condition` in that it flags conditions have constant behavior at runtime, since that is highly correlated with developer error. Specifically it checks for:

* `&&` and `||` logical expressions where the left hand side is either always truthy or always falsy 
* `??` logical expressions where the left hand side is either always nullish or always non-nullish
* `==`, `===`, `!==` and `!=` binary expressions where one side is either `null`, `undefined`, `true` or `false`, and the outcome of the comparison can be determined statically.
 
#### Is there anything you'd like reviewers to focus on?

The code ended up begin quite verbose with comments, which I personally found helpful during development, but may not be idiomatic in this code base. I'm happy to tone them down if desired.

At its heart this rule boils down to a "best effort". It understands many semantics of the JavaScript language, but it must draw the line somewhere, and those lines are necessarily somewhat arbitrary. I feel pretty good about the lines I ended up setting given the [TSC Summary](https://github.com/eslint/eslint/issues/13752#issuecomment-729125654), but I'm open to feedback.

## Impact and Representative Examples

I ran this rule on our large monorepo and it **uncovered ~450 existing errors**. Of those, about one third looked like changes which were intended to be temporary behavior changes, or hard coded feature flags:

```JavaScript
const enableFeature = false && shouldEnableFeature();
```

Personally I think these would merit an `/* eslint-disable-next-line */` comment to clarify the motivation, but I could also see us wanting to add an option to permit literal `true` or `false` on the left hand side of `||` or `&&`.

Below are some representative errors it uncovered, anonymized and reduced to their pure form:

```jsx
value === (0 || undefined || null)

!value !== null

someFunction() != null ?? false

x + b ?? 0

String(someNumber) ?? 'DEFAULT_VALUE'

typeof someMemberExpression !== undefined

switch (status) {
  case 'A' || 'B' || 'C' || 'D':
    // some code
    break;
  // [...]
}

+value ?? 0

!!someArr.find(predicate) ?? false
```

## Rule Name

**Edit** @mdjermanovic [pointed out](https://github.com/eslint/eslint/pull/15296#issuecomment-979329860) that the TSC had proposed a better name, but I missed the comment.

~~Is `no-constant-binary-expression` the right name for this rule given that it also applies to the left hand side of logical expressions? *I'd rather not re-litigate, but my instinct tells me that the logical expression checks should live in `no-constant-condition` and this new rule should flag any comparison binary expression whose result is provably static.*~~

### CODE REVIEW

File: lib/rules/no-constant-binary-operand.js
Comment (bradzacher) [inline lib/rules/no-constant-binary-operand.js:66]: `delete` can be easily reasoned about I think.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/delete#return_value

In non-strict mode it returns true unless the property being deleted is non-configurable.
In strict mode it always returns true and throws in the latter case.

So the logic could be:
```ts
if (context.getScope().isStrict) {
  // delete will always returns true or throw
  return true;
} else {
  // not strict mode
  if (node.type === 'MemberExpression') {
    // might return false if the property being deleted is non-configurable
    return false;
  }
  return true;
}
```

Comment (bradzacher) [reply lib/rules/no-constant-binary-operand.js:66]: For `+`, `-`, `~` - yeah they get a bit involved to handle - but it's possible to statically handle some cases.

Might be a good idea to write this as a TODO - in case anyone wants to solve this in future?

File: lib/rules/no-constant-binary-operand.js
Comment (bradzacher) [inline lib/rules/no-constant-binary-operand.js:198]: Note 1 - this will crash for `{...a}`.
Note 2 - `{...a}` is not constant as `a` might have one of the keys.

File: lib/rules/no-constant-binary-operand.js
Comment (bradzacher) [inline lib/rules/no-constant-binary-operand.js:207]: I'm not sure I understand this.
Won't `[0]` be converted to `'0'` (and similarly for `[1]`) - which are constant strings.
Am I misunderstanding the conversion?

Comment (captbaritone) [reply lib/rules/no-constant-binary-operand.js:207]: Yeah, an array with statically declared elements would be fine (that's what the TODO is trying to call out), but `[x]` cannot be reasoned about statically since:

```
[1] == true
=> true
[0] == true
=> false
```

File: lib/rules/no-constant-binary-operand.js
Comment (captbaritone) [reply lib/rules/no-constant-binary-operand.js]: Fixed

File: lib/rules/no-constant-binary-operand.js
Comment (captbaritone) [reply lib/rules/no-constant-binary-operand.js]: I've updated the comment to be a little more clear.

File: conf/eslint-recommended.js
Comment (bmish) [inline conf/eslint-recommended.js:23]: Rules shouldn't be added as `recommended`. We can only add to `recommended` in a major release.

File: lib/rules/no-constant-binary-operand.js
Comment (bmish) [inline lib/rules/no-constant-binary-operand.js:384]: There's a new comment that should be added here per https://github.com/eslint/eslint/pull/15291:

```suggestion
/** @type {import('eslint').Rule.RuleModule} */
module.exports = {
```

**Comment by j-f1** [comment]
> Is `no-constant-binary-operand` the right name for this rule given that it also applies to the left hand side of logical expressions?

When first reading through this PR, I thought this rule prohibited constant operands in binary expressions (such as the 1 in `x + 1`). I assume at least some other people will feel the same way and may skip over it in a list because of that interpretation.

File: lib/rules/no-constant-binary-operand.js
Comment (captbaritone) [reply lib/rules/no-constant-binary-operand.js:387]: Looks like it ended up needing to be `/** @type {import('../shared/types').Rule} */`.

File: lib/rules/no-constant-binary-operand.js
Comment (captbaritone) [reply lib/rules/no-constant-binary-operand.js:66]: `delete` is pretty rare, and supporting it will require passing scope (or at least `isStrict`) into the helpers. I think I'll punt on supporting `delete` for now, but if we end up with some other reason to include scope or other contextual data in the helpers, I'll add support at that point.

**Comment by captbaritone** [comment]
I believe all previous feedback has been addressed.

**Review by Swapnull — COMMENTED** [review]
Hey, I hope this is ok, but I just noticed you are spelling `always` and `short-circuiting` wrong in a few comments. It seems to be fairly consistent, so a find a replace should sort it :)

**Comment by mdjermanovic** [comment]
> Is `no-constant-binary-operand` the right name for this rule given that it also applies to the left hand side of logical expressions?

Per https://github.com/eslint/eslint/issues/13752#issuecomment-730765050, the name should be `no-constant-binary-expression`.

**Comment by captbaritone** [comment]
> the name should be no-constant-binary-expression

Ah thanks. Will fix.

File: docs/rules/no-constant-binary-expression.md
Comment (nzakas) [inline docs/rules/no-constant-binary-expression.md:1]: I don’t think short-circuiting applies to all conditions in the rule, so maybe 

```suggestion
# disallow expressions where the operation doesn’t affect the value (no-constant-binary-expression)
```

File: docs/rules/no-constant-binary-expression.md
Comment (nzakas) [inline docs/rules/no-constant-binary-expression.md:3]:
```suggestion
Comparisons which will always evaluate to true or false and logical expressions (`||`, `&&`, `||`) which either always short circuit or never short circuit are both likely indications of programmer error.
```

File: docs/rules/no-constant-binary-expression.md
Comment (nzakas) [inline docs/rules/no-constant-binary-expression.md:5]: This is a good spot for a simple example.

File: docs/rules/no-constant-binary-expression.md
Comment (nzakas) [inline docs/rules/no-constant-binary-expression.md:9]:
```suggestion
This rule identifies `==` and `===` comparisons which, based on the semantics of the JavaScript language, will always evaluate to `true` or `false`.
```

File: docs/rules/no-constant-binary-expression.md
Comment (nzakas) [inline docs/rules/no-constant-binary-expression.md:11]:
```suggestion
It also identifies `||`, `&&` and `??` logical expressions which will either always or never short circuit.
```

File: docs/rules/no-constant-condition.md
Comment (nzakas) [inline docs/rules/no-constant-condition.md:78]: Our convention is to have a related rules section at the bottom of documentation. See https://eslint.org/docs/rules/no-ternary#related-rules

File: lib/rules/no-constant-binary-expression.js
Comment (nzakas) [inline lib/rules/no-constant-binary-expression.js:51]: Are these TODOs to be completed before merging?

File: lib/rules/no-constant-binary-expression.js
Comment (nzakas) [inline lib/rules/no-constant-binary-expression.js:168]: Can you describe here what patterns are checked?

File: lib/rules/no-constant-binary-expression.js
Comment (nzakas) [inline lib/rules/no-constant-binary-expression.js:391]: Make sure this matches the title on the docs page

File: tests/lib/rules/no-constant-binary-expression.js
Comment (nzakas) [inline tests/lib/rules/no-constant-binary-expression.js:21]: Can you move this down to where the variable is used? We generally have valid cases first and invalid second in the test files.

**Review by nzakas — CHANGES_REQUESTED** [review]

File: lib/rules/no-constant-binary-expression.js
Comment (captbaritone) [reply lib/rules/no-constant-binary-expression.js:51]: These TODOs represent ways in which the rule could technically be taken further, but felt like they passed the point of diminishing returns in my estimation. I wasn't planning to complete these as part of this PR, but if reviewers think any of them seem worth doing, I'd be happy to add them.

My intention was that they would clarify places where I recognized that further inference was technically possible, but had opted not to do that work while also leaving a hint in case someone in the future wanted to consider add it.

Comment (captbaritone) [reply lib/rules/no-constant-binary-expression.js:51]: I'm happy to reword these to be more stylistically consistent with the rest of the project, just let me know what would work.

**Comment by captbaritone** [comment]
In #15357 I've proposed expanding this rule to also detect comparison to always-new objects: `foo === {}`, `foo === []`, `foo === (() => {})` etc.

I'll wait for affirmative approval on that issue before updating this PR to include that logic.

File: lib/rules/no-constant-binary-expression.js
Comment (nzakas) [reply lib/rules/no-constant-binary-expression.js:51]: Yeah, I’d maybe say something like “Possible future direction if needed:” just to make it clear that there’s no tech debt here, just an enhancement opportunity.

Comment (captbaritone) [reply lib/rules/no-constant-binary-expression.js:51]: Done.

File: lib/rules/no-constant-binary-expression.js
Comment (mdjermanovic) [inline lib/rules/no-constant-binary-expression.js:24]: Thoughts on reusing [`function isConstant`](https://github.com/eslint/eslint/blob/d041f345cdf0306f15faa2f305fe1d21ef137eb1/lib/rules/no-constant-condition.js#L123-L131) from `no-constant-condition`?

It looks like both functions are meant to determine the same thing, which is whether an expression has constant truthiness, so perhaps we could have only one function in a shared place (`ast-utils`) and use it from both rules.

That would catch more cases in `no-constant-binary-expression`. For example, the current version of `no-constant-binary-expression` doesn't report any of these:

```js
-1 || foo; 

1 + 2 || f; 

`a${1}b` === "c" || foo;

a || true || b;
```

While the `no-constant-condition` rule reports same expressions as constant conditions:

```js
-1 ? foo : bar; // no-constant-condition error

1 + 2 ? foo : bar; // no-constant-condition error

`a${1}b` === "c" ? foo : bar; // no-constant-condition error

a || true || b ? foo : bar; // no-constant-condition error
```

(we could add the `Boolean` and `undefined` checks from this implementation to the shared function)

Comment (nzakas) [reply lib/rules/no-constant-binary-expression.js:24]: We shouldn’t have one rule reach into another rule to use a function. If these two functions do indeed intend to determine the same thing (@captbaritone can comment on that), then moving it into ast-utils makes sense.

Comment (captbaritone) [reply lib/rules/no-constant-binary-expression.js:24]: I think this is a great point. `isConstant` goes to greater effort to detect cases where you can determine that truthiness is constant based based on having constant subexpressions and is therefore it is able to catch more issues.

I agree that it would make sense to extract that into a helper function and use it in both places. But this also makes me think that `hasConstantNullishness` should probably be equally aggressive in trying to understand the impact of subexpressions. I'll need to take a little more time to see how/if there's any way for those two cases to share some of this logic.

**Comment by nzakas** [comment]
@captbaritone just checking in to see if you’re still working on this?

**Comment by captbaritone** [comment]
> @captbaritone just checking in to see if you’re still working on this?

Yes! Just been busy.

**Comment by nzakas** [comment]
No problem. I just like to check in when I see PRs lose momentum. No hurry!

File: lib/rules/no-constant-binary-expression.js
Comment (captbaritone) [reply lib/rules/no-constant-binary-expression.js:24]: I'm still trying to make sense of `isConstant` and what exactly we are checking when `inBooleanPosition` is `false`. I finally realized that at least part of my confusion was due to the fact that there are some edge cases which it does not handle correctly. I've reported those as https://github.com/eslint/eslint/issues/15467.

I think that currently `isConstant` with `inBooleanPosition` set to `false` is _supposed_ to check if a value has constant truthiness, _even when that value is being first coerced to some other type_ (string/number/...). I'm not sure that's a function which is feasible to write correctly. Instead, I suspect we will want individual functions that handle the `string` and `number` cases, or something along those lines.

File: lib/rules/no-constant-binary-expression.js
Comment (mdjermanovic) [inline lib/rules/no-constant-binary-expression.js:472]: Would it be possible to also check `==` and `!=`?

Something like:

```js
} else if (operator === "==" || operator === "!=") {
    if (isAlwaysNew(left) && isAlwaysNew(right)) {
        // report
    }
}
```

File: lib/rules/no-constant-binary-expression.js
Comment (captbaritone) [reply lib/rules/no-constant-binary-expression.js:24]: With https://github.com/eslint/eslint/pull/15486 merged, I think it _does_ make sense to extract `isConstant` (and therefore its helper functions) into a utility shareable by multiple rules. Should I do that as part of this PR, or a separate PR?

File: lib/rules/no-constant-binary-expression.js
Comment (captbaritone) [reply lib/rules/no-constant-binary-expression.js:472]: This would work for this specific case, and would be easy to implement. I think it's probably worth adding.

If we wanted to generalize this approach, we would need to define some helpers which try to statically determine the type (object/string/number/null/undefined) of a given expression. We could then combine the knowledge of the type of each side of the comparison in order to determine if their comparison would always be static. However, this is significantly more involved, so I recommend we don't take that on, or at the very least defer it to a separate project.

Comment (nzakas) [reply lib/rules/no-constant-binary-expression.js:472]: Yeah, I don’t think it’s worth trying to add type detection here (that would need an RFC and would probably be better off as part of eslint-scope). Let’s just stick with the simple solution for this PR

Comment (captbaritone) [reply lib/rules/no-constant-binary-expression.js:492]: I've updated the code to cover this case.

**Comment by captbaritone** [comment]
I've removed the code that tries to account for object expressions that include `toValue`, `toString`, etc. This keeps us consistent with the [feedback](https://github.com/eslint/eslint/pull/15486#discussion_r778461384) I got in my changes to `no-constant-condition`.

File: lib/rules/no-constant-binary-expression.js
Comment (captbaritone) [reply lib/rules/no-constant-binary-expression.js:23]: @nzakas Thoughts on how/if to go about moving `isConstant` from the `no-constant-condition` rule (and a few functions it depends upon) into a util?

Comment (mdjermanovic) [reply lib/rules/no-constant-binary-expression.js:23]: It's fine to extract `isConstant` (and its helper functions) to `rules/ast-utils.js` as a part of this PR.

Comment (mdjermanovic) [reply lib/rules/no-constant-binary-expression.js:23]: > It's fine to extract `isConstant` (and its helper functions) to `rules/ast-utils.js` as a part of this PR.

Also, since there are plenty of tests for `no-constant-condition` rule, which basically just calls `isConstant`, I think separate unit tests for the extracted `astUtils.isConstant` are not necessary at the moment, so we can just copy & paste the code.

**Comment by captbaritone** [comment]
Finally got around to trying to share `isConstant` from `is-constant-condition`. In that process, I noticed a few cases that it misses which my implementation catches. I've opened #15613 to close those gaps before trying to move/share it.

**Comment by captbaritone** [comment]
Now that #15613 has been merged, I've updated this PR to include extracting `isConstant`, and its associated helper functions, from `no-constant-condition` into `ast-utils` and using it here. I've done this as a second commit to hopefully make the impact of that decision a bit more clear.

The main impact is that a lot more functions need to accept `scope` as an additional parameter.

**Comment by captbaritone** [comment]
(Test failure looks like a flaky test to me)

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

File: lib/rules/no-constant-binary-expression.js
Comment (mdjermanovic) [inline lib/rules/no-constant-binary-expression.js:60]: We can use [`astUtils.isLogicalAssignmentOperator`](https://github.com/eslint/eslint/blob/18f5e05bce10503186989d81ca484abb185a2c9d/lib/rules/utils/ast-utils.js#L755) here.

**Comment by mdjermanovic** [comment]
> I've done this as a second commit to hopefully make the impact of that decision a bit more clear.

Feel free to always make changes to PRs by adding commits, when possible. It's easier for reviewers, as described in https://eslint.org/docs/developer-guide/contributing/pull-requests#updating-the-code

File: lib/rules/no-constant-binary-expression.js
Comment (captbaritone) [reply lib/rules/no-constant-binary-expression.js:317]: Oh good catch, I didn't realize that! We could keep it but restrict it to non-shadowed built-ins. I would love to still be able to catch boxed primitive bugs like:

```
x = new Boolean(someVar) === true;
```

Just checking for the primitives: `String`, `Number`, `Boolean` might be enough, but we could also consider a large set of builtins like `Set`, `Array`, `Map`, `WeakMap` etc.

Comment (captbaritone) [reply lib/rules/no-constant-binary-expression.js:317]: I looked into this a bit more, and some of the non-primitive constructors (eg. `Promise`) don't work correctly with `isReferenceToGlobalVariable`. A non-shadowed construction `new Promise` does not have a resolved value:

```javascript
x === new Promise()
```

```javascript
const reference = scope.references.find(ref => ref.identifier === node);
// reference.resolved === null
```

I can look into the scope implementation to learn more in the next day or so, but I figured I'd flag here in case you had any ideas in the mean time.

Comment (mdjermanovic) [reply lib/rules/no-constant-binary-expression.js:317]: > I looked into this a bit more, and some of the non-primitive constructors (eg. `Promise`) don't work correctly with `isReferenceToGlobalVariable`. A non-shadowed construction `new Promise` does not have a resolved value:

Did you enable an es6+ environment? In test cases, it would look like this:

https://github.com/eslint/eslint/blob/8bb527660ffddc122204fb5bf59076687aaa78b8/tests/lib/rules/no-obj-calls.js#L32-L35

ESLint by default adds es5 globals only, so by default there is no `Promise` global and references to it would be unresolved.

Comment (mdjermanovic) [reply lib/rules/no-constant-binary-expression.js:317]: > We could keep it but restrict it to non-shadowed built-ins.

Makes sense to me. We could check whether or not the global is an ES built-in by using [`globals.builtin`](https://github.com/eslint/eslint/blob/8bb527660ffddc122204fb5bf59076687aaa78b8/lib/rules/no-extend-native.js#L56) like in the `no-extend-native` rule.

**Comment by captbaritone** [comment]
This is ready for another round of review.

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

**Comment by captbaritone** [comment]
@mdjermanovic Thanks for the detailed review with examples of code that would behave give incorrect results. Made it super easy to address the feedback and add the missing tests cases.

Updated and ready for another round.

File: lib/rules/no-constant-binary-expression.js
Comment (mdjermanovic) [inline lib/rules/no-constant-binary-expression.js:381]: I can't figure out why this check logically belongs to `findBinaryExpressionConstantOperand`, while the similar one for `===` and `!==` is done separately? Their messages are also different in the current version of this PR:

```js
/* eslint no-constant-binary-expression: "error" */

[a] === [b]; // Unexpected comparison to newly constructed object. These two values can never be equal.

[a] == [b]; // Unexpected constant binary expression. Compares constantly with the left-hand side of the `==`.

```

File: lib/rules/no-constant-binary-expression.js
Comment (mdjermanovic) [inline lib/rules/no-constant-binary-expression.js:165]: This could be wrong when the array expression has spread elements:

```js
/* eslint no-constant-binary-expression: "error" */

function test1(arr) {
    const eq = [...arr, 1] == true; // reported as a constant binary expression

    console.log(eq);
}

test1([]); // logs `true`
test1([1]); // logs `false`

```

Maybe we could return `true` only if there are 0 elements or > 1 elements that are not SpreadElement?

File: lib/rules/no-constant-binary-expression.js
Comment (captbaritone) [reply lib/rules/no-constant-binary-expression.js]: Good point. I've pulled this out and given it its own message.

File: lib/rules/no-constant-binary-expression.js
Comment (captbaritone) [reply lib/rules/no-constant-binary-expression.js]: We had been using `isNullOrUndefined` from `ast-utils.js` which did not perform a scope check. Interestingly, this meant that I was unable to write a regression test for this change because `isNullOr Undefined` would detect the `undefined` side as undefined, and then `hasConstantNullishness` would return true for `true`.

I decided to fork `isNullOrUndefined` and create a local version which does perform the scope check. If we truly consider redefined `undefined` and edge case we need to support, we should consider a followup to add `scope` as an additional argument to `isNullOrUndefined` from `ast-utils.js`.

**Comment by captbaritone** [comment]
@mdjermanovic Again, great feedback thanks! Ready for another pass.

File: lib/rules/no-constant-binary-expression.js
Comment (mdjermanovic) [reply lib/rules/no-constant-binary-expression.js]: Good idea, we can update `astUtils.isNullOrUndefined` and the rules that are using it in a followup PR, and then switch this rule to it as well.

File: docs/src/rules/no-constant-binary-expression.md
Comment (mdjermanovic) [inline docs/src/rules/no-constant-binary-expression.md:61]:
```suggestion
const objIsEmpty = Object.keys(someObj).length === 0;
```

Same as the previous.

File: docs/src/rules/no-constant-binary-expression.md
Comment (mdjermanovic) [inline docs/src/rules/no-constant-binary-expression.md:2]:
```suggestion

Disallows expressions where the operation doesn't affect the value.

```

We usually add the short description (basically the same text as `meta.docs.description`) as the first paragraph in the docs.

**Comment by captbaritone** [comment]
@mdjermanovic  Feedback addressed.

File: lib/rules/no-constant-binary-expression.js
Comment (mdjermanovic) [inline lib/rules/no-constant-binary-expression.js:444]:
```suggestion
            bothAlwaysNew: "Unexpected comparison of two newly constructed objects. These two values can never be equal."
```

**Review by mdjermanovic — COMMENTED** [review]
Just one small typo to fix, and then LGTM.

**Comment by captbaritone** [comment]
@mdjermanovic Fixed

**Review by mdjermanovic — APPROVED** [review]
Great work, thanks!

I'll leave this open for @nzakas to confirm his feedback has been addressed.

File: lib/rules/utils/ast-utils.js
Comment (nzakas) [inline lib/rules/utils/ast-utils.js:884]:
```suggestion
 * @returns {boolean} true when node's truthiness is constant
```

File: docs/src/rules/no-constant-binary-expression.md
Comment (nzakas) [inline docs/src/rules/no-constant-binary-expression.md:5]:
```suggestion
Comparisons which will always evaluate to true or false and logical expressions (`||`, `&&`, `??`) which either always short-circuit or never short-circuit are both likely indications of programmer error.
```

File: docs/src/rules/no-constant-binary-expression.md
Comment (nzakas) [inline docs/src/rules/no-constant-binary-expression.md:17]:
```suggestion
Additionally, this rule detects comparisons to newly constructed objects/arrays/functions/etc. In JavaScript, where objects are compared by reference, a newly constructed object can _never_ `===` any other value. This can be surprising for programmers coming from languages where objects are compared by value.
```

File: docs/src/rules/no-constant-binary-expression.md
Comment (nzakas) [inline docs/src/rules/no-constant-binary-expression.md:30]:
```suggestion
It also identifies `||`, `&&` and `??` logical expressions which will either always or never short-circuit.
```

**Review by nzakas — APPROVED** [review]
LGTM. I just left a few suggestions for cleaning up the docs and JSDoc.

**Comment by captbaritone** [comment]
@nzakas Thanks! I've merged committed them all.

**Comment by mdjermanovic** [comment]
I'll update the doc file for https://github.com/eslint/eslint/pull/15782 in a follow-up PR.

**Comment by mdjermanovic** [comment]
@captbaritone thanks for contributing!

**Pull request merged**
